### ISSUE TITLE
Add Nextcloud support

### ISSUE THREAD

**Issue by ezzra** [LycheeOrg/Lychee#580]
Similar to dropbox, there could be an option to import photos from Nextcloud installations.

**Comment by ildyria** [comment]
I started having a quick look. Sadly I won't have the time to dive deeper.

Aside from getting the `Storage::put()` to work this will require a lot of refactoring. The reason being that currently we:

1. upload the picture
2. place it in the said folder and delete the tmp_file 
3. Generate the exif data, dimensions etc
4. Generate the Photo object (record in the DB)
5. Generate the thumbs.
6. Save

Basically this requires to move the thumbnail generation before the Generation of the photo object.
This will require significant rewrite of the code which I don't have the time at the moment.

I am currently doing a huge refactoring of the code, which hopefully will simplify this part.

**Comment by tokenwizard** [comment]
Rather than trying to directly integrate with NextCloud, could Lychee be configured to look for photos in a local folder? That path could be pointed to the NC Data folder that contains the Photos and then when you upload a photo to NC, it could be detected and added to the Lychee Library? 

PhotoView uses this type of functionality, so you can continue using NC to manage your files and PhotoView gives you a "separate" gallery to view them in a nice way.

**Comment by d7415** [comment]
> could Lychee be configured to look for photos in a local folder?

Essentially, [yes](https://lycheeorg.github.io/docs/faq.html#can-i-set-up-lychee-to-watch-a-folder-for-new-images-and-automatically-add-them-to-albums).

**Comment by mariomare22** [comment]
> Rather than trying to directly integrate with NextCloud, could Lychee be configured to look for photos in a local folder? That path could be pointed to the NC Data folder that contains the Photos and then when you upload a photo to NC, it could be detected and added to the Lychee Library? 
> 
> PhotoView uses this type of functionality, so you can continue using NC to manage your files and PhotoView gives you a "separate" gallery to view them in a nice way. 

I like that, but i was aiming on some UI integration over maybe WebDAV and also have the auto scan manageable from the UI itself and not Cron...

**Comment by d7415** [comment]
> manageable from the UI itself

Not automated, but just to mention that there's also "Import from server" in the UI that does the same as the CLI command.

**Comment by justadri** [comment]
> Basically this requires to move the thumbnail generation before the Generation of the photo object. This will require significant rewrite of the code which I don't have the time at the moment.
> 
> I am currently doing a huge refactoring of the code, which hopefully will simplify this part.

What's the status of this issue today? Has the photo upload refactoring been done such that a folder scan/sync implementation would be possible without rewriting the core code? I'm willing to help with the dev work, I just want to get an idea of the amount of effort involved and what, if any, work has been done so far.

**Comment by sunjam** [comment]
What about the long running [Lycheesync](https://github.com/GustavePate/lycheesync)?  It has been around forever and serves the purpose of adding an arbitrary directory to lychee.  Coded in Python.

**Comment by d7415** [comment]
> What about the long running [Lycheesync](https://github.com/GustavePate/lycheesync)? It has been around forever and serves the purpose of adding an arbitrary directory to lychee. Coded in Python.

That only works with Lychee v3
